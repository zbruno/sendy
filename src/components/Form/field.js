import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons';

// eslint-disable-next-line react/prefer-stateless-function
class Field extends React.Component {
  static propTypes = {
    disabled: PropTypes.bool,
    hasError: PropTypes.bool,
    label: PropTypes.string,
    labelFor: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    type: PropTypes.string,
  };

  static defaultProps = {
    disabled: false,
    hasError: false,
    label: '',
    placeholder: '',
    type: 'text',
  };

  render() {
    const {
      disabled,
      hasError,
      label,
      labelFor,
      placeholder,
      value,
      onChange,
      type,
    } = this.props;

    const labelClasses = cx({
      'block text-gray-700 text-sm font-bold mb-2': true,
    });
    const inputClasses = cx({
      'shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline': true,
      'disabled:cursor-not-allowed disabled:bg-gray-200': true,
      'border-red-500': hasError,
    });
    const errorIconClasses = cx({
      'opacity-0 absolute inset-0 left-auto h-100 flex justify-center items-center pointer-events-none mr-2': true,
      'opacity-100': hasError,
    });

    return (
      <div className="flex flex-col mb-6">
        <label className={labelClasses} htmlFor={labelFor}>
          {label}
        </label>
        <div className="relative">
          <div className={errorIconClasses}>
            <FontAwesomeIcon
              icon={faTimesCircle}
              className="fill-current text-red-500"
              size="lg"
              fixedWidth
            />
          </div>
          <input
            className={inputClasses}
            id={labelFor}
            type={type}
            placeholder={placeholder}
            value={value}
            disabled={disabled}
            onChange={onChange}
          />
        </div>
      </div>
    );
  }
}

export default Field;
