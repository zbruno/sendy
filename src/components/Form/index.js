import React from 'react';
import PropTypes from 'prop-types';

import Field from './field';
import Card from '../Card';

class Form extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    title: PropTypes.string.isRequired,
  };

  static Field = props => (
    <Field {...props} />
  );

  render() {
    const { children, title } = this.props;
    return (
      <Card title={title}>
        <form>
          {children}
        </form>
      </Card>
    );
  }
}

export default Form;
