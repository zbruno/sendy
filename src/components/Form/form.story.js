import React, { useState } from 'react';

import Form from '.';

export default {
  title: 'Form',
  component: Form,
  parameters: {
    jest: [],
  },
};

export const form = () => {
  const [email, setEmail, password, setPassword] = useState('');

  return (
    <Form title="A Form!">
      <Form.Field
        disabled={false}
        hasError={false}
        label="Email address"
        labelFor="email"
        placeholder=""
        onChange={event => {
          setEmail(event.target.value);
        }}
        value={email}
        type="email"
      />
      <Form.Field
        disabled={false}
        hasError={false}
        label="Password"
        labelFor="password"
        placeholder=""
        onChange={event => {
          setPassword(event.target.value);
        }}
        value={password}
        type="password"
      />
    </Form>
  );
};
