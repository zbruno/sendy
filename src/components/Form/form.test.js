import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Form from './index';

describe('<Form />', () => {
  describe('UI Snapshot', () => {
    test('renders correctly', () => {
      const labelText = 'Test Label';
      const labelFor = 'for-input-id';
      const component = renderer.create(
        <Form title={labelText}>
          <Form.Field
            label={labelText}
            labelFor={labelFor}
            value=""
            onChange={() => {}}
          />
        </Form>,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('onChange callback', () => {
    test('should be invoked onChange', () => {
      const spy = jest.fn();
      const form = mount(
        <Form.Field
          label=""
          labelFor=""
          value=""
          onChange={spy}
        />,
      );

      form.find('input').simulate('change');
      expect(spy).toHaveBeenCalled();
    });
  });
});
