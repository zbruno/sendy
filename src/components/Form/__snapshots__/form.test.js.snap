// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> UI Snapshot renders correctly 1`] = `
<div
  className="flex flex-col break-words bg-white border border-2 rounded shadow-md"
>
  <div
    className="font-semibold bg-purple-500 text-white py-3 px-6 mb-0 rounded-t"
  >
    Test Label
  </div>
  <div
    className="w-full py-4 px-6"
  >
    <form>
      <div
        className="flex flex-col mb-6"
      >
        <label
          className="block text-gray-700 text-sm font-bold mb-2"
          htmlFor="for-input-id"
        >
          Test Label
        </label>
        <div
          className="relative"
        >
          <div
            className="opacity-0 absolute inset-0 left-auto h-100 flex justify-center items-center pointer-events-none mr-2"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-times-circle fa-w-16 fa-fw fa-lg fill-current text-red-500"
              data-icon="times-circle"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
          <input
            className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline disabled:cursor-not-allowed disabled:bg-gray-200"
            disabled={false}
            id="for-input-id"
            onChange={[Function]}
            placeholder=""
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;
